core/src/main/java/org/apache/iceberg/rest/HTTPHeaders.java

```diff
@@ -0,0 +1,109 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.rest;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.immutables.value.Value;
+
+/**
+ * Represents a group of HTTP headers.
+ *
+ * <p>Header name comparison in this class is always case-insensitive, in accordance with RFC 2616.
+ *
+ * <p>This class exposes methods to convert to and from different representations such as maps and
+ * multimap, for easier access and manipulation – especially when dealing with multiple headers with
+ * the same name.
+ */
+@Value.Style(depluralize = true)
+@Value.Immutable
+@SuppressWarnings({"ImmutablesStyle", "SafeLoggingPropagation"})
+public interface HTTPHeaders {
+
+  HTTPHeaders EMPTY = of();
+
+  /** Returns all the header entries in this group. */
+  List<HTTPHeader> entries();
+
+  /** Returns all the entries in this group for the given name (case-insensitive). */
+  default List<HTTPHeader> entries(String name) {
+    return entries().stream()
+        .filter(header -> header.name().equalsIgnoreCase(name))
+        .collect(Collectors.toList());
+  }
+
+  /** Returns whether this group contains an entry with the given name (case-insensitive). */
+  default boolean contains(String name) {
+    return entries().stream().anyMatch(header -> header.name().equalsIgnoreCase(name));
+  }
+
+  /**
+   * Adds the given header to the current group if no entry with the same name is already present.
+   * Returns a new instance with the added header, or the current instance if the header is already
+   * present.
+   */
+  default HTTPHeaders withHeaderIfAbsent(HTTPHeader header) {
+    Preconditions.checkNotNull(header, "header");
+    return contains(header.name())
+        ? this
+        : ImmutableHTTPHeaders.builder().from(this).addEntry(header).build();
+  }
+
+  /**
+   * Adds the given headers to the current group if no entries with same names are already present.
+   * Returns a new instance with the added headers, or the current instance if all headers are
+   * already present.
+   */
+  default HTTPHeaders withHeaderIfAbsent(HTTPHeaders headers) {
+    Preconditions.checkNotNull(headers, "headers");
+    List<HTTPHeader> newHeaders =
+        headers.entries().stream().filter(e -> !contains(e.name())).collect(Collectors.toList());
+    return newHeaders.isEmpty()
+        ? this
+        : ImmutableHTTPHeaders.builder().from(this).addAllEntries(newHeaders).build();
+  }
+
+  static HTTPHeaders of(HTTPHeader... headers) {
+    return ImmutableHTTPHeaders.builder().addEntries(headers).build();
+  }
+
+  /** Represents an HTTP header as a name-value pair. */
+  @Value.Style(redactedMask = "****", depluralize = true)
+  @Value.Immutable
+  @SuppressWarnings({"ImmutablesStyle", "SafeLoggingPropagation"})
+  interface HTTPHeader {
+
+    String name();
+
+    @Value.Redacted
```

Comment:
Does this mean that we redact all header values?  Maybe I don't fully understand the Immutable behavior here, but that's a bit problematic because we probably only want to redact sensitive headers (like `AUTHORIZATION`).

Reply:
It's not possible to redact just some headers and not all of them unfortunately.

Reply:
And why is it "problematic" to redact all header values? Iceberg won't ever call `HTTPHeaders.toString()` on purpose, so this feature is here only to prevent accidentally printing a header value to the console.

Reply:
I would assume that the apache http client has logging as well, but I guess it would have been translated to their headers, so I'm good with it here if we don't plan to log.